onnx_trt_backend.cpp

```diff
@@ -0,0 +1,749 @@
+        case nvonnxparser::ErrorCode::kINVALID_GRAPH:
+        case nvonnxparser::ErrorCode::kINVALID_NODE:
+          return ONNXIFI_STATUS_INVALID_MODEL;
+        case nvonnxparser::ErrorCode::kUNSUPPORTED_NODE:
+        case nvonnxparser::ErrorCode::kUNSUPPORTED_GRAPH:
+          return ONNXIFI_STATUS_UNSUPPORTED_OPERATOR;
+        default:
+          return ONNXIFI_STATUS_INTERNAL_ERROR;
+        }
+      }
+    }
+
+    return ONNXIFI_STATUS_SUCCESS;
+  }
+
+  nvinfer1::ICudaEngine *buildCudaEngine() {
+    return trt_builder_->buildCudaEngine(*trt_network_);
+  }
+
+  size_t max_batch_size() const { return max_batch_size_; }
+
+private:
+  TRT_Logger trt_logger_;
+  cudaStream_t stream_;
+  std::shared_ptr<nvinfer1::IBuilder> trt_builder_{nullptr};
+  std::shared_ptr<nvinfer1::INetworkDefinition> trt_network_{nullptr};
+  std::shared_ptr<nvonnxparser::IParser> parser_{nullptr};
+  // TODO: configerable max batch size
+  int device_id_{0};
+  size_t max_batch_size_{128};
+};
+
+class GraphRep {
+public:
+  GraphRep(OnnxTensorRTBackendRep *backendrep)
+      : device_id_(backendrep->device_id()),
+        max_batch_size_(backendrep->max_batch_size()),
+        stream_(backendrep->stream()) {
+    if (cudaSetDevice(device_id_) != cudaSuccess) {
+      throw std::runtime_error("Cannot set CUDA device");
+    }
+    trt_engine_ = infer_object(backendrep->buildCudaEngine());
+    max_batch_size_ = backendrep->max_batch_size();
+  }
+
+  ~GraphRep() { ClearDeviceBuffers(); }
+
+  onnxStatus InitIO(uint32_t inputsCount,
+                    const onnxTensorDescriptor *inputDescriptors,
+                    uint32_t outputsCount,
+                    const onnxTensorDescriptor *outputDescriptors);
+
+  onnxStatus Run();
+
+  cudaStream_t stream() const { return stream_; }
+
+private:
+  void ClearDeviceBuffers();
+
+  onnxStatus CheckAndBindTensor(const nvinfer1::Dims &dims,
+                                const onnxTensorDescriptor &tensor);
+
+  std::shared_ptr<nvinfer1::ICudaEngine> trt_engine_{nullptr};
+  std::shared_ptr<nvinfer1::IExecutionContext> trt_executor_{nullptr};
+  std::vector<void *> bindings_;
+  std::unordered_map<std::string, const onnxTensorDescriptor *> input_map_;
+  std::unordered_map<std::string, const onnxTensorDescriptor *> output_map_;
+  std::unordered_map<std::string, void *> device_buffers_;
+  int device_id_{0};
+  size_t max_batch_size_{0};
+  size_t batch_size_{0};
+  cudaStream_t stream_;
+};
+
+void GraphRep::ClearDeviceBuffers() {
+  for (auto kv : device_buffers_) {
+    cudaFree(kv.second);
+  }
+  device_buffers_.clear();
+}
+
+onnxStatus GraphRep::CheckAndBindTensor(const nvinfer1::Dims &dims,
+                                        const onnxTensorDescriptor &tensor) {
+  // Check memory type
+  if (tensor.memoryType != ONNXIFI_MEMORY_TYPE_CPU &&
+      tensor.memoryType != ONNXIFI_MEMORY_TYPE_CUDA_BUFFER) {
+  std::cerr << tensor.memoryType << std::endl;
+    return ONNXIFI_STATUS_INVALID_DATATYPE;
```

Comment:
Oops, I should add a proper error code for this situation

Reply:
Oops, I should remove the `std::cerr` line above.